### DISCUSSION TITLE
Switching tabs noticeable slow

### DISCUSSION THREAD

**Discussion by medwatt** [qtile/qtile#5006] [Q&A]
I am trying out qtile from i3 and liking it so far, but there is a noticable lag when switching between tabs. Is this normal?

https://github.com/user-attachments/assets/f722aff5-a294-4b6d-8ba5-329c8fde8ef9

**Comment by ervinpopescu** [comment]
I assume this is X11, and I haven't even tried TreeTab so far, but I'd bet that switching windows like this will indeed show a lag. 

The issue might or might not be python, it could also be Xorg running slower on your system or it could be qtile doing the wrong thing (I'd say less likely). I literally have no idea which one of these it is. 

Which layout commands are you using in the video to switch between windows?

**Reply by medwatt** [reply]
This is X11. The issue is the same with max layout.
I am using the following keys to move:

```
    Key([mod], "left", lazy.layout.left(), desc="Move focus to left"),
    Key([mod], "right", lazy.layout.right(), desc="Move focus to right"),
    Key([mod], "down", lazy.layout.down(), desc="Move focus down"),
    Key([mod], "up", lazy.layout.up(), desc="Move focus up"),
```

**Reply by saviola777** [reply]
I did some tests, for me the slight delay when switching to the inkscape window happens irrespective of layout and both when switching groups and windows within a group. It looks like it just takes a while to render the inkscape window upon switching to it. There is no lag when switching away from Inkscape and I haven't noticed a similar lag with any other application.

**Comment by jrwrigh** [comment]
I'm on X11 and use TreeTab regularly. I've never seen this kind of lag before. Could possibly a compositor issue? Or GPU. I'm running integrated graphics.

**Reply by medwatt** [reply]
I also have intel integrated graphics. I am using picom.

**Reply by medwatt** [reply]
Maybe, try it with inkscape. It shows the worse lag.

**Reply by jrwrigh** [reply]
Ok, with inkscape I'm seeing some lag, but not nearly as bad as what you're showing. The lag is also only there when trying to make it appear, not when moving to a non-inkscape window.

**Reply by ervinpopescu** [reply]
> Ok, with inkscape I'm seeing some lag, but not nearly as bad as what you're showing. The lag is also only there when trying to make it appear, not when moving to a non-inkscape window.

Same results here with picom on (i.e. no lag). With picom off, does it still have lag?

**Reply by jrwrigh** [reply]
I'm still seeing the lag with and without picom.

**Reply by ervinpopescu** [reply]
Sorry, I edited my comment, it wasn't clear that I see no lag.

**Reply by medwatt** [reply]
The blinking is very irritating. I may have to move back to i3.

**Reply by jrwrigh** [reply]
Yeah, I don't have much of an idea as to what could be causing this. Since it's application dependent, I would've suspected an X11 issue maybe, but if i3 doesn't have those issues, seems like it might have something to do with qtile implementation.

@m-col @tych0 @elParaguayo Tagging y'all since y'all might have a better idea of what could be causing this and how to fix it.

**Reply by tych0** [reply]
I am not really familiar with treetab's code, but my guess is that it's unmapping the old window, and then mapping the new one. My guess is that if we switched it to focus and draw the new window before unmapping the old one, the flickering wouldn't happen.

**Reply by jrwrigh** [reply]
Perhaps it's some sort of hide/show latency that i3 masks by always having "stacked" windows (windows that would occupy the same space) still be "displayed" but hidden somehow?

---
Here's a gif demonstrating the hysteresis in the latency/blinking. 
- Going from a terminal to Inkscape results in a significant delay, 
- Going from Inkscape to a terminal still has a delay, but is much less significant
- Going from terminal to terminal is instantaneous

![qtile_inkscape_treetab_latency](https://github.com/user-attachments/assets/7f1b5d28-8a03-465d-b068-caf8876a224c)

**Reply by jrwrigh** [reply]
> My guess is that if we switched it to focus and draw the new window before unmapping the old one, the flickering wouldn't happen.

Flickering would go away, but the latency would remain, yes? I think that's probably fine, since the latency is mostly noticeable for the flickering.

**Reply by tych0** [reply]
We do a bunch of synchronous things in x11's hide() (changing the window mask to ignore events... twice [we should perhaps coalesce this]). Those synchronous calls both happen before we try to map the window, and are probably quite slow. So I think even latency would improve if we mapped and then unmapped.

**Reply by medwatt** [reply]
To clarify, this issue does not just happen with treetab, it happens with all layouts that display one full window on the screen at once. For example, treetab, max, stack(with one column), etc. However, the degree of laginess is different. It is most noticeable with the max layout, even worse than the original video I posted, at least on my system.

Furthermore, the issues does not just happen with inkscape. It happens with many other programs. It's justs that it's more pronouced with inkscape. For example, open two pdfs in okular in max layout, zoom out the page so that you have a lot of black background. Then try to switch between the two pdfs. You will see wallpaper as you do so.

I just tried with awesomewm, and it is even smoother than i3. So, I guess this points to the issue being with qtile. To be honest, coming from i3, qtile doesn't feel as smooth when resizing things or changing stuff on the screen. It is small, but diffinitely noticeable, and may point to unoptimized code somewhere in the loop.

**Reply by jrwrigh** [reply]
Going through the "backtrace":
- `TreeTab.next()` https://github.com/qtile/qtile/blob/1d9ddfca4cc2f65e327ba9b250e2a20b0669fa85/libqtile/layout/tree.py#L567
- `Group.focus()` https://github.com/qtile/qtile/blob/1d9ddfca4cc2f65e327ba9b250e2a20b0669fa85/libqtile/layout/tree.py#L575
  - calls `TreeTab.focus()` https://github.com/qtile/qtile/blob/1d9ddfca4cc2f65e327ba9b250e2a20b0669fa85/libqtile/group.py#L232
  - `TreeTab.focus()` just sets the `_focused` member to the window https://github.com/qtile/qtile/blob/1d9ddfca4cc2f65e327ba9b250e2a20b0669fa85/libqtile/layout/tree.py#L405-L406
- `Group.layout_all()` (run from `Group.focus()`) https://github.com/qtile/qtile/blob/1d9ddfca4cc2f65e327ba9b250e2a20b0669fa85/libqtile/group.py#L234
- `TreeTab.layout()` https://github.com/qtile/qtile/blob/1d9ddfca4cc2f65e327ba9b250e2a20b0669fa85/libqtile/group.py#L160
- `Layout.layout()` https://github.com/qtile/qtile/blob/1d9ddfca4cc2f65e327ba9b250e2a20b0669fa85/libqtile/layout/tree.py#L758
- Loops through windows and calls `TreeTab.configure()`https://github.com/qtile/qtile/blob/1d9ddfca4cc2f65e327ba9b250e2a20b0669fa85/libqtile/layout/base.py#L59-L61
- `TreeTab.configure()` determines which windows to show or hide https://github.com/qtile/qtile/blob/1d9ddfca4cc2f65e327ba9b250e2a20b0669fa85/libqtile/layout/tree.py#L483-L490
  - This will then hide/unhide the windows

@tych0 Is there a "need" to call `Layout.layout()`, or can this be done custom by TreeTab? I'm thinking something like this should work fine:
```diff
diff --git i/libqtile/layout/tree.py w/libqtile/layout/tree.py
index df6e5d63..948f9c55 100644
--- i/libqtile/layout/tree.py
+++ w/libqtile/layout/tree.py
@@ -755,7 +755,12 @@ class TreeTab(Layout):
         else:
             panel, body = screen_rect.hsplit(self.panel_width)
         self._resize_panel(panel)
-        Layout.layout(self, windows, body)
+        if self._nodes and self._focused:
+            self.configure(self._focused, body)
+        for win in windows:
+            if win is self._focused:
+                continue
+            self.configure(win, body)

     def _resize_panel(self, screen_rect):
         if self._panel:
```

The only "issue" I see is that if `client.unhide()` (inside `TreeTab.configure()`) isn't blocking then we'll still end up having the same issue. But I'd imagine X11 makes that a blocking call, so shouldn't be an issue?

**Reply by elParaguayo** [reply]
I don't see any lag on X11 with windows in fullscreen (e.g. Max layout), including with Inkscape.

**Comment by tych0** [comment]
> Is there a "need" to call Layout.layout(), or can this be done custom by TreeTab? I'm thinking something like this should work fine:

I think the idea is that this code is shared between all layouts because they basically all need to do the same thing. You are right that `unhide()` is also synchronous, so it should be done before any laying out is done.

Given that is the case, is there any reason your optimization can't be applied to Layout's `layout()`?

Maybe that one can't, but it seems like we can certainly skip withdrawn/minimized/etc. windows rather than synchronously repainting them all on every layout call...

**Reply by jrwrigh** [reply]
> Given that is the case, is there any reason your optimization can't be applied to Layout's `layout()`?

Well the current optimization utilizes the `_focused` member in TreeTab, so that's not portable. Maybe there's a way to make an equivalent portable across layouts, but I'm not familiar with all of them.

> Maybe that one can't, but it seems like we can certainly skip withdrawn/minimized/etc. windows rather than synchronously repainting them all on every layout call...

Maybe, but in this case we need to call `configure` on a window that _is_ withdrawn/minimized. Regardless, the synchronous repainting, while unnecessary, doesn't appear to result in any significant lag. At least on my system.

**Comment by tych0** [comment]
> Regardless, the synchronous repainting, while unnecessary, doesn't appear to result in any significant lag. At least on my system.

Yeah, I don't see any lag at all on mine either, it was just a guess. I suppose before we start optimizing, we should figure out where the lag is.

**Reply by jrwrigh** [reply]
Well, I feel like the lag isn't as much a concern as the flickering. Given the lag is application dependent, I doubt there's anything major that Qtile could do on it's side to help that.

**Reply by tych0** [reply]
Yeah, I'm lumping lag in with flickering, I think they're the same problem: we are unmapping before we map. If we can do it in the other order in the general case for all layouts, that would be best.

Does your patch above fix the flickering for treetab?

**Reply by jrwrigh** [reply]
> Does your patch above fix the flickering for treetab?

That would presume I've tested it. :sweat_smile: 

I'll try and see where Arch installed Qtile and see if the patch fixes the issue.

**Reply by tych0** [reply]
Ok; if it does, I think we should figure out how to make the optimization for all layouts, not just treetab. there is `window.has_focus()`, perhaps we can use that in the generalized implementation.

**Reply by jrwrigh** [reply]
Oddly, it doesn't appear to have any effect. Even after restarting my computer, still has the flickering effect.

**Reply by tych0** [reply]
Huh, that is strange. I wonder if the actual unmapping of the window is happening before we even call `layout()`? Like some rogue code somewhere that is doing an unmap via `.hide()` and then calling `layout()`? (Maybe `.hide()` should not unmap, and instead wait for `layout()` to be called to redraw everything in the right order?)

**Comment by telephon** [comment]
We see it on macOS as well. (https://github.com/supercollider/supercollider)

**Comment by medwatt** [comment]
@tych0, any update on this issue?
